### PR TITLE
Hcmpre 4161 airflow trigger

### PR DESCRIPTION
**Pull request by hruthvikl-egov**
Extends the airflow-trigger-service so the UI can show accurate, per-run report status and avoid redundant regenerations. Adds two new query APIs, an append-only REPORTS_METADATA lifecycle
  (with a self-tracking DB migration), row-count/time estimates, a re-trigger cooldown, dynamic locale, and a timezone-correctness fix for custom date ranges. No breaking changes to the
  existing trigger flow.

 New APIs
  - POST /airflow-trigger-api/api/reports-check-existing — before triggering a CUSTOM (date-range) report, the UI checks whether one already exists for the exact reportRange. Returns the
  existing run, or exists:false when a completed run is past its cooldown / when nothing exists (so the UI's "trigger if none" path is unchanged). Failed runs allow immediate retry;
  in-progress ones stay blocked.
  - POST /airflow-trigger-api/api/reports-in-progress — lists in-flight runs so the UI can render live progress.

  Status lifecycle on REPORTS_METADATA (append-only)
  - The table is now one row per status event (TRIGGERED → POD_STARTED → … → REPORT_COMPLETED), not just terminal rows; the persister writes with ON CONFLICT (eventId) DO NOTHING.
  - Responses are enriched server-side (_enrich_status_row): isTerminal, isFailed, reportTimeSeconds (report-script runtime), processingTimeSeconds (trigger→completion), elapsedSeconds,
  secondsSinceTriggered, expectedRows, expectedGenerationTimeSeconds, and a live progressPercent.
  - search_reports_metadata filters to REPORT_COMPLETED (plus legacy rows with a real fileStoreId) to preserve the original "one row per successful report" shape.

  Row/time estimates
  - On trigger, expected rows and expected generation time are computed via dashboard-analytics getChartV2 (*_estimate charts), stamped once and carried through every subsequent status event
  — used to drive the live progress bar.

Re-trigger cooldown
  - A completed CUSTOM report for the same range is blocked from re-generation for CUSTOM_REPORT_RETRY_COOLDOWN_SECONDS (default 3600), returning retryAvailableInSeconds/Minutes. Per-report
  overrides (incl. "never") via CUSTOM_REPORT_RETRY_COOLDOWN_OVERRIDES. Failed/expired → immediate retry.

  Dynamic locale
  - TriggerRequest accepts an explicit locale and RequestInfo; locale resolves body locale → RequestInfo.msgId (<ts>|<locale>) → default, matching excel-ingestion.

  Central-instance / schema handling
  - Reads/writes resolve the table as {tenantId}.REPORTS_METADATA when IS_CENTRAL_INSTANCE_ENABLED=true, else REPORTS_METADATA under search_path = DB_SCHEMA (first of DB_SCHEMAS). Keeps the
  trigger-service, persister, and migration on the same schema.

  Timezone fix (_parse_custom_range)
  - Removed the UTC conversion of the custom start/end. The epoch is unchanged, but converting shifted the displayed reportRange back a day (e.g. 13th IST → 12th UTC). Now the selected day
  is preserved — matching the DAG's compute_range CUSTOM branch.

  DB migration (migrate.py)
  - Self-tracking runner (DB_SCHEMAS → CREATE SCHEMA + search_path, default public); each migration runs once per schema.
  - Adds the REPORTS_METADATA lifecycle columns (status/eventId/timings/estimates/etc.) and a unique index on eventId so the persister's ON CONFLICT (eventId) works.

 Backward compatibility
  - Existing trigger endpoint unchanged. Legacy REPORTS_METADATA rows (pre-status) still surface in reports-metadata when they have a fileStoreId. New columns are additive.

New config / env (set in health-campaign-devops):

    DB_SCHEMAS  :                            Schema used for the migration and for reads.
                                            e.g. "health" (unified) or "demo" (central-instance demo).

    IS_CENTRAL_INSTANCE_ENABLED    :         "true"  -> table resolved as {tenantId}.REPORTS_METADATA
                                            "false" -> plain REPORTS_METADATA under search_path=DB_SCHEMA

    EGOV_DASHBOARD_ANALYTICS_HOST    :       dashboard-analytics host; used for expected-row
                                            estimates via getChartV2 (*_estimate charts).

    CUSTOM_REPORT_RETRY_COOLDOWN_SECONDS  :   Cooldown before an identical custom report can be
                                            re-triggered. Default 3600 (1 hour).

    CUSTOM_REPORT_RETRY_COOLDOWN_OVERRIDES :  Optional. JSON of per-report overrides
                                            (value in seconds, or "NEVER" to permanently block).

## Summary by CodeRabbit

- **New Features**
  - Added richer report lifecycle tracking, including status history, in-progress reports, progress percentages, elapsed times, and failure indicators.
  - Added tenant-local date filtering for report metadata.
  - Added safeguards against duplicate or overly frequent CUSTOM report regenerations.
  - Enhanced CUSTOM report triggers with locale, timing, and expected-row details.
  - Added configurable database schema support.

- **Bug Fixes**
  - Improved Airflow authentication recovery and error reporting.
  - Added safer trigger handling with status recording and retry protection.
  - Improved validation and handling of report status and lifecycle requests.

### CODE REVIEW

**Review by Taniya-eGov — APPROVED** [review]

**Review by prerna-egov — APPROVED** [review]

**Review by naveen-egov — COMMENTED** [review]
🟠 1. Blocking DB I/O inside async endpoints — blocks the event loop.
All four new endpoints are async def but call synchronous psycopg2 (_get_db_conn, cur.execute, fetchall). _insert_triggered_on_ui_rows (async) and _fetch_expected_generation_time_seconds (sync, awaited-path in trigger_dag) also do blocking DB work with no threadpool. Under concurrent load this serializes requests on the single event loop. The pre-existing endpoint had this, but the PR adds four more. Fix: either make these plain def (FastAPI runs sync handlers in a threadpool) or wrap DB calls in starlette.concurrency.run_in_threadpool / asyncio.to_thread.

🟠 2. reports_in_progress: magic statusOrder < 40 for "in flight".
WHERE rn = 1 AND statusOrder < 40 AND status <> 'SKIPPED'
40 is undocumented and coupled to the external STATUS_ORDER (in kafka_status.py, not in this repo) — if any in-progress stage (POD_STARTED, REPORT_GENERATION_STARTED, …) has order ≥ 40, live runs silently vanish from the UI. It's also inconsistent with the same file's set-based TERMINAL_STATUSES/FAILED_STATUSES. Prefer deriving in-flight from status NOT IN (<terminal set>) (or at minimum a named constant + comment justifying 40), so the two definitions can't drift.

🟡 3. createdTime TIMESTAMP → BIGINT is a full-table rewrite.
ALTER COLUMN ... TYPE BIGINT USING (...) takes an ACCESS EXCLUSIVE lock and rewrites every row — on a now-append-only, growing REPORTS_METADATA this can block writes during deploy, and there's no down-migration (framework has no rollback). Fine at current size; worth a heads-up / off-peak deploy as the table grows.

🟡 4. No tests (0 added for +902 lines).
Several pure, edge-case-heavy functions are ideal unit targets: _parse_custom_range/_compute_custom_report_range (timezone preservation, bad-format 400), _get_retry_cooldown_seconds ("NEVER"/null/int override), _enrich_status_row (terminal vs live progressPercent, null guards), and the cooldown-remaining math in check_existing_custom_report. This is the biggest quality gap.

🟡 5. Raw exception leaked to clients.
raise HTTPException(status_code=500, detail=str(e)) in all four endpoints returns DB/SQL error text to the caller (info disclosure). You already logger.exception(...) server-side — return a generic message to the client instead.

🔵 6. Minor efficiency / DRY.
- _insert_triggered_on_ui_rows opens a new DB connection per campaign inside the loop — open once outside, reuse.
- The ROW_NUMBER() OVER (PARTITION BY … ORDER BY statusOrder DESC, eventTimestampMs DESC) "latest per run" pattern is copy-pasted across three endpoints — factor into a helper.
- trigger_dag awaits getChartV2 + blocking DB per CUSTOM campaign sequentially (usually few, so low impact).

**Review by naveen-egov — APPROVED** [review]

File: utilities/airflow-trigger-service/api/main.py [hunk 1]
```diff
@@ -297,15 +890,258 @@ async def search_reports_metadata(req: ReportsMetadataRequest):
...
+    # magic `statusOrder < 40` threshold, so it can't silently drift from kafka_status.py's
+    # ordering if a new in-progress stage is added. NULL-status legacy rows fall out of
+    # NOT IN, which is correct - they were never real in-flight runs.
+    terminal_statuses = tuple(sorted(TERMINAL_STATUSES))
+    terminal_placeholders = ",".join(["%s"] * len(terminal_statuses))
+    query = (
+        _latest_per_run_subquery(table, where_clause)
+        + f" WHERE rn = 1 AND status NOT IN ({terminal_placeholders}) ORDER BY eventTimestampMs DESC"
+    )
+    params = params + list(terminal_statuses)
```
[the remark this answers is not in the document]
Comment (hruthvikl-egov) [reply utilities/airflow-trigger-service/api/main.py]: Not in scope

**Comment by hruthvikl-egov** [comment]
> Caution
> 
> Some comments are outside the diff and can’t be posted inline due to platform limitations.
> 
> ⚠️ Outside diff range comments (1)
> > utilities/airflow-trigger-service/api/main.py (1)> `900-909`: _🩺 Stability & Availability_ | _🟠 Major_ | _⚡ Quick win_
> > > **Close database resources when the query fails.**
> > > `cur` and `conn` close only after `fetchall()` succeeds. A repeated database error can leave connections open and exhaust the database connection limit.
> > > 
> > > Proposed fix
> > > ```diff
> > > +    conn = None
> > > +    cur = None
> > >      try:
> > >          conn = _get_db_conn()
> > >          cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
> > >          cur.execute(query, params)
> > >          rows = cur.fetchall()
> > > -        cur.close()
> > > -        conn.close()
> > >      except Exception:
> > >          logger.exception("search_reports_metadata query failed")
> > >          raise HTTPException(status_code=500, detail="Failed to fetch report metadata")
> > > +    finally:
> > > +        if cur is not None:
> > > +            cur.close()
> > > +        if conn is not None:
> > > +            conn.close()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 
> > > 🤖 Prompt for AI Agents
> > > ```
> > > Verify each finding against current code. Fix only still-valid issues, skip the
> > > rest with a brief reason, keep changes minimal, and validate.
> > > 
> > > In `@utilities/airflow-trigger-service/api/main.py` around lines 900 - 909, Update
> > > the database query flow around _get_db_conn and the cursor operations so cur and
> > > conn are closed when execute or fetchall raises, using guaranteed cleanup while
> > > preserving the existing logging and HTTPException behavior.
> > > ```
> 
> 🤖 Prompt for all review comments with AI agents
> ```
> Verify each finding against current code. Fix only still-valid issues, skip the
> rest with a brief reason, keep changes minimal, and validate.
> 
> Outside diff comments:
> In `@utilities/airflow-trigger-service/api/main.py`:
> - Around line 900-909: Update the database query flow around _get_db_conn and
> the cursor operations so cur and conn are closed when execute or fetchall
> raises, using guaranteed cleanup while preserving the existing logging and
> HTTPException behavior.
> ```
> 
> ℹ️ Review info
> ⚙️ Run configuration
> **Configuration used**: Path: .coderabbit.yaml
> 
> **Review profile**: ASSERTIVE
> 
> **Plan**: Pro Plus
> 
> **Run ID**: `6279861f-ae8f-4e80-a260-c1a83d0a8a5a`
> 
> 📥 Commits
> Reviewing files that changed from the base of the PR and between [1dbd349](https://github.com/egovernments/health-campaign-services/commit/1dbd349c940853bad28ff9bc76d7071d4fe2f7cd) and [f38a7d8](https://github.com/egovernments/health-campaign-services/commit/f38a7d88268ed70e5b9131c5b63712463f3a57aa).
> 
> 📒 Files selected for processing (2)
> * `utilities/airflow-trigger-service/api/main.py`
> * `utilities/airflow-trigger-service/api/test_main.py`

**Review by kavi-egov — APPROVED** [review]
